v2/pkg/engine/jsonschema/variables_schema.go

```diff
@@ -0,0 +1,461 @@
+		varSchema.Nullable = false
+	}
+
+	// Add variable to schema
+	v.schema.Properties[varName] = varSchema
+}
+
+// processOperationTypeRef processes a type reference from the operation document
+func (v *VariablesSchemaBuilder) processOperationTypeRef(typeRef int) *JsonSchema {
+	switch v.operationDocument.Types[typeRef].TypeKind {
+	case ast.TypeKindNonNull:
+		ofType := v.operationDocument.Types[typeRef].OfType
+		schema := v.processOperationTypeRef(ofType)
+		if schema == nil {
+			return nil
+		}
+		// Non-null types are not nullable
+		schema.Nullable = false
+		return schema
+
+	case ast.TypeKindList:
+		ofType := v.operationDocument.Types[typeRef].OfType
+		itemSchema := v.processOperationTypeRef(ofType)
+		if itemSchema == nil {
+			return nil
+		}
+		// If we're not in a non-null context, list is nullable
+		schema := NewArraySchema(itemSchema)
+		schema.Nullable = true
+		return schema
+
+	case ast.TypeKindNamed:
+		typeName := v.operationDocument.TypeNameString(typeRef)
+		schema := v.processTypeByName(typeName)
+		if schema != nil {
+			// If we're not in a non-null context, named type is nullable
+			schema.Nullable = true
+		}
+		return schema
+	}
+
+	return nil
+}
+
+// processTypeByName processes a type by its name, looking it up in the definition document
+func (v *VariablesSchemaBuilder) processTypeByName(typeName string) *JsonSchema {
+	// Handle built-in scalars
+	switch typeName {
+	case "String":
+		return NewStringSchema()
+	case "Int":
+		return NewIntegerSchema()
+	case "Float":
+		return NewNumberSchema()
+	case "Boolean":
+		return NewBooleanSchema()
+	case "ID":
+		return NewStringSchema()
+	}
+
+	// For custom types, look up in the definition document
+	node, exists := v.definitionDocument.Index.FirstNodeByNameStr(typeName)
+	if !exists {
+		v.report.AddInternalError(fmt.Errorf("type %s is not defined", typeName))
+		return NewObjectSchema()
+	}
+
+	// Check recursion depth for complex types that could be recursive
+	if node.Kind == ast.NodeKindEnumTypeDefinition || node.Kind == ast.NodeKindInputObjectTypeDefinition {
+		currentDepth, exists := v.recursionTracker[typeName]
+		if exists {
+			// We've seen this type before
+			currentDepth++
+			v.recursionTracker[typeName] = currentDepth
+
+			// If we've hit our recursion limit, return nil to signal field removal
+			if currentDepth > v.maxRecursionDepth {
+				return nil
+			}
+		} else {
+			// First time seeing this type
+			v.recursionTracker[typeName] = 1
+		}
+
+		// Defer the cleanup of the recursion tracker
+		defer func() {
+			if depth, ok := v.recursionTracker[typeName]; ok && depth > 1 {
+				v.recursionTracker[typeName]--
+			} else {
+				delete(v.recursionTracker, typeName)
+			}
+		}()
+	}
+
+	switch node.Kind {
+	case ast.NodeKindEnumTypeDefinition:
+		return v.processEnumType(node)
+
+	case ast.NodeKindInputObjectTypeDefinition:
+		return v.processInputObjectType(node)
+
+	case ast.NodeKindScalarTypeDefinition:
+		return NewStringSchema()
```

Comment:
Custom Scalar should be any, not string. I think any in JSON Schema can be represented as `{}` but I'm not 100% sure.

Reply:
Good one. Empty object can be used for it.